docs/executor_plugins.md

```diff
@@ -0,0 +1,140 @@
+# Executor Plugins
+
+## Types
+
+### Template Executor
+
+There is only one type of executor plugin, one that runs custom ("plugin") templates, e.g. for non-pod tasks such as
+Tekton builds or Spark jobs.
+
+## A Simple Python Plugin
+
+Lets make a Python plugin that prints "hello" each time the workflow is operated on.
+
+We need the following:
+
+1. [Plugins enabled](plugins.md) .
+2. A HTTP server that will be run as a sidecar to the main container responds to RPC HTTP requests from the executor.
+3. Configuration so the controller can discover the plugin.
+
+We'll need to create a script that starts a HTTP server:
+
+```python
+import json
+from http.server import BaseHTTPRequestHandler, HTTPServer
+
+
+class Plugin(BaseHTTPRequestHandler):
+
+    def args(self):
+        return json.loads(self.rfile.read(int(self.headers.get('Content-Length'))))
+
+    def reply(self, reply):
+        self.send_response(200)
+        self.end_headers()
+        self.wfile.write(json.dumps(reply).encode("UTF-8"))
+
+    def unsupported(self):
+        self.send_response(404)
+        self.end_headers()
+
+    def do_POST(self):
+        if self.path == '/api/v1/template.execute':
+            args = self.args()
+            if 'hello' in args['template'].get('plugin', {}):
+                self.reply({'node': {'phase': 'Succeeded', 'message': 'Hello template!'}})
+            else:
+                self.reply({})
+        else:
+            self.unsupported()
+
+
+if __name__ == '__main__':
+    httpd = HTTPServer(('', 4355), Plugin)
```

Comment:
How about setting the port number as an ENV var so that it can be injected by the executor/controller at runtime? Not sure if there is any actual need for this though...

Reply:
I like that idea a lot. Makes it so there's a single source of truth for the port number. The env var could be something like `PLUGIN_PORT`?

Reply:
Actually, I feel like making the containerPort config optional would be even nicer. If it's absent, pick one and inject the YAML in the built plugin. If it's present, use the given port.

Reply:
+1 good idea

Reply:
in fact, really good idea, solves all port collision problems

Reply:
I've just tried to implement this:

* It means we need to modify the user provided container to add the port.
* We want to support user-provided port.

So I think we leave this for MVP.

Reply:
Open ticket?